execution-engine/src/hcall/common.rs

```diff
@@ -1003,196 +361,124 @@ pub(crate) enum EntrySignature {
-        self.request_shutdown(client_id);
-        self.is_able_to_shutdown()
-    }
+/// These are return codes that the host passes back to the Veracruz WASM program
+/// when something goes wrong with a host-call.  Any error is assumed to be
+/// recoverable by the WASM program, if it cares to, and are distinct from execution engine
+/// errors which are akin to kernel panics and are always fatal.
+///
+/// Note that both the host and any Veracruz program need to agree on how these
+/// errors are encoded.
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+pub enum EngineReturnCode {
+    /// The H-call completed successfully.
+    Success,
+    /// Generic failure: no more-specific information about the cause of the
+    /// error can be given.
+    Generic,
+    /// The H-call failed because an index was passed that exceeded the number
+    /// of data sources.
+    DataSourceCount,
+    /// The H-call failed because it was passed a buffer whose size did not
+    /// match the size of the data source.
+    DataSourceSize,
+    /// The H-call failed because it was passed bad inputs.
+    BadInput,
+    /// The H-call failed because an index was passed that exceeded the number
+    /// of stream sources.
+    StreamSourceCount,
+    /// The H-call failed because it was passed a buffer whose size did not
+    /// match the size of the stream source.
+    StreamSourceSize,
+    /// The H-call failed because it was passed bad streams.
+    BadStream,
+    /// The H-call failed because it was passed a buffer whose size did not
+    /// match the size of the previous result.
+    PreviousResultSize,
+    /// An internal invariant was violated (i.e. we are morally "panicking").
+    InvariantFailed,
+    /// The H-call failed because a result had already previously been written.
+    ResultAlreadyWritten,
+    /// The H-call failed because the platform service backing it is not
+    /// available on this platform.
+    ServiceUnavailable,
 }
 
 ////////////////////////////////////////////////////////////////////////////////
-// Trait implementations
+// Trait implementations.
 ////////////////////////////////////////////////////////////////////////////////
 
-/// Serialize a `LifecycleState` value to a `u8`.
-///
-/// This is needed for responding to the Veracruz protocol "query enclave state"
-/// request from a computation principal.  The current state is encoded as a
-/// `u8` and forwarded to the requestor.
-impl From<LifecycleState> for u8 {
-    fn from(state: LifecycleState) -> Self {
-        match state {
-            LifecycleState::Initial => 0,
-            LifecycleState::DataSourcesLoading => 1,
-            LifecycleState::StreamSourcesLoading => 2,
-            LifecycleState::ReadyToExecute => 3,
-            LifecycleState::FinishedExecuting => 4,
-            LifecycleState::Error => 5,
+/// Pretty printing for `EngineReturnCode`.
+impl Display for EngineReturnCode {
```

Comment:
Why not use `err_derive` here, like in a lot of the other error types?

Reply:
I renamed this type from an old type. It is use in the `Ok(_)` case as the return code from wasm program.